File: leetcode/editor/cn/160_相交链表_test.py
import unittest
from typing import Optional

from model.list_node import ListNode


class TestSolution(unittest.TestCase):

    def setUp(self) -> None:
        self.solution = Solution()

    def test_givenNormal_when_thenSuccess(self):
        n3 = ListNode(3)
        h1 = ListNode(1, ListNode(2, n3))
        h2 = ListNode(4, n3)
        node = self.solution.getIntersectionNode(h1, h2)
        self.assertEqual(n3, node)


# leetcode submit region begin(Prohibit modification and deletion)
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, x):
#         self.val = x
#         self.next = None

class Solution:
    def getIntersectionNode(self, headA: ListNode, headB: ListNode) -> Optional[ListNode]:
        pa, pb = headA, headB
        while pa != pb:
            pa = pa.next if pa else headB
            pb = pb.next if pb else headA
        return pa


# leetcode submit region end(Prohibit modification and deletion)


if __name__ == '__main__':
    unittest.main()

File: leetcode/editor/cn/160_相交链表_test.go
package leetcode

import (
	"github.com/stretchr/testify/assert"
	"testing"
)

func Test_givenNormal_when160_thenSuccess(t *testing.T) {
	n3 := &ListNode{3, nil}
	h1 := &ListNode{1, &ListNode{2, n3}}
	h2 := &ListNode{4, n3}
	assert.Equal(t, n3, getIntersectionNode(h1, h2), "shouldEqual")
}

//leetcode submit region begin(Prohibit modification and deletion)
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func getIntersectionNode(headA, headB *ListNode) *ListNode {
	if headA == nil || headB == nil {
		return nil
	}
	pa, pb := headA, headB
	// nil == nil is true
	for pa != pb {
		if pa != nil {
			pa = pa.Next
		} else {
			pa = headB
		}
		if pb != nil {
			pb = pb.Next
		} else {
			pb = headA
		}
	}
	return pa
}

//leetcode submit region end(Prohibit modification and deletion)
